kvbc/src/kvbc_adapter/v4blockchain/blocks_adder_adapter.cpp

```diff
@@ -0,0 +1,121 @@
+                        << concordUtils::bufferToHex(to_be_written_block_header_key.c_str(),
+                                                     to_be_written_block_header_key.size()));
+
+  if (cat_itr != updates.categoryUpdates().kv.cend()) {
+    const auto &kvs = std::get<kvbc::categorization::VersionedInput>(cat_itr->second).kv;
+    auto key_itr = kvs.find(to_be_written_block_header_key);
+    if (key_itr != kvs.cend()) {
+      // Eth block
+      header = BlockHeader::deserialize(key_itr->second.data);
+      found_existing_header = true;
+    }
+  }
+
+  if (not found_existing_header) {
+    // Reconfig block; !Eth block;
+    header.data.number = to_be_written_block_num;
+  }
+
+  writeParentHash(to_be_written_block_num, header.data.parent_hash);
+
+  // write block number to block header mapping into updates
+  auto serialized_header = header.serialize();
+  updates.addCategoryIfNotExisting<kvbc::categorization::VersionedInput>(
+      concord::kvbc::categorization::kExecutionPrivateCategory);
+  updates.appendKeyValue<kvbc::categorization::VersionedUpdates>(
+      concord::kvbc::categorization::kExecutionPrivateCategory,
+      std::move(to_be_written_block_header_key),
+      kvbc::categorization::VersionedUpdates::Value{serialized_header, true});
+
+  // write block num to block hash mapping in updates
+  auto header_hash = header.hash();
+  kvbc::categorization::VersionedUpdates::Value val{};
+  val.data.append(std::begin(header_hash), std::end(header_hash));
+  val.stale_on_update = true;
+  auto block_num_as_key_to_block_hash =
+      BlockHeader::blockNumAsKeyToBlockHash((uint8_t *)&to_be_written_block_num, sizeof(to_be_written_block_num));
+  updates.appendKeyValue<kvbc::categorization::VersionedUpdates>(
+      concord::kvbc::categorization::kExecutionPrivateCategory,
+      std::move(block_num_as_key_to_block_hash),
+      std::move(val));
+
+  // write block header hash to block number mapping in updates
+  auto block_hash_as_key_to_block_number =
+      BlockHeader::blockHashAsKeyToBlockNum((uint8_t *)&header_hash, sizeof(concord::kvbc::uint256be_t));
+  updates.appendKeyValue<kvbc::categorization::VersionedUpdates>(
+      concord::kvbc::categorization::kExecutionPrivateCategory,
+      std::move(block_hash_as_key_to_block_number),
+      kvbc::categorization::VersionedUpdates::Value{concordUtils::toBigEndianStringBuffer(to_be_written_block_num),
+                                                    true});
+
+  LOG_DEBUG(GL,
+            "Block number " << to_be_written_block_num << " header size " << serialized_header.size()
+                            << " and its header hash "
+                            << concordUtils::bufferToHex(header_hash.data(), header_hash.size()));
+}
+
+BlockId BlocksAdderAdapter::add(concord::kvbc::categorization::Updates &&updates) {
+  if (bftEngine::ReplicaConfig::instance().extraHeader == true) {
+    addHeader(std::forward<concord::kvbc::categorization::Updates>(updates));
```

Comment:
I think it will be more appropriate for the API off addHeader to accept the updates as a reference and not an rvalue i.e. to omit the std::forward as the addHeader adds data to the updates but doesn't move them.

Reply:
Addressed. Thanks.